### PR TITLE
Add integrated Secure Enclave block signing for nodeos

### PR DESCRIPTION
**Pull request by spoonincode**
## Change Description

`nodeos` has supported signing blocks with the Secure Enclave since 1.2 by using the KEOSD signature provider. That is, `nodeos` would ask `keosd` to sign blocks with the SecureEnclave wallet. However, that isn’t entirely secure because `keosd` can’t really validate requests to sign are from `nodeos` and not something else (even an unsigned process). This change adds a new signature provider to `nodeos` that can sign directly with the Secure Enclave.

To manage Secure Enclave keys available to `nodeos` a new application `nodeos-sectl` must be used. With it, one can create, list, and remove Secure Enclave keys available to `nodeos`. Use the `-h` option on this application for usage instructions. But as a simple example creating a key is as easy as
```
% nodeos-sectl.app/Contents/MacOS/nodeos-sectl -c
PUB_R1_6f5K98nbNTEeBgW7kvstof7oqniabozHWSz3Jwszr3ZdMLrCeY
```

Once a key has been created via `nodeos-sectl`, `nodeos` can be configured to sign with it by adding an appropriate signature provider. For example,
```
signature-provider = PUB_R1_6f5K98nbNTEeBgW7kvstof7oqniabozHWSz3Jwszr3ZdMLrCeY=SE:
```

Using the Secure Enclave requires EOSIO to be signed properly. Prior to this change, all applications in eosio signed for Secure Enclave usage would share the same keychain group meaning they would share the same keys. This was most interesting/helpful in eosio 1.1 because at the time it was still possible to run the wallet plugin inside of `nodeos`. The shared keychain group meant the wallet plugin in `nodeos` would see the same keys that the wallet plugin in `keosd` would see.

However, going forward this is absolutely not the behavior we want. We want different keys in `nodeos` for signing blocks vs keys in `keosd` for signing transactions.

To accomplish this, the cmake variable `MAS_KEYCHAIN_GROUP` is now optional and discouraged. If not specified `nodeos` will end up with a keychain group of `MAS_BASE_APPIDnodeos` and `keosd` will end up with a keychain group of `MAS_BASE_APPIDkeosd`; they won’t be able to see keys created by each other.

Previous users of Secure Enclave keys will need to continue specifying `MAS_KEYCHAIN_GROUP` to see the keys created in prior versions using that keychain group.

## Consensus Changes
- [ ] Consensus Changes

## API Changes
- [ ] API Changes

## Documentation Additions
- [x] Documentation Additions

Any previous documentation on configuring Secure Enclave builds needs to be updated; I'm not sure we ever documented this though because it's rather hairy.

### CODE REVIEW

File: programs/nodeos-sectl/main.cpp [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+   try {
+      bpo::store(bpo::parse_command_line(argc, argv, cli), varmap);
+      bpo::notify(varmap);
+   }
+   catch(fc::exception& e) {
+      elog("${e}", ("e", e.to_detail_string()));
+      return 1;
+   }
+
+   if((!list && !create && key_to_delete == fc::crypto::public_key()) || help) {
```
[the remark this answers is not in the document]
Comment (spoonincode) [reply programs/nodeos-sectl/main.cpp]: Should they? This wasn't clearly obvious to me. Would someone ever want to do `nodeos-sectl -d PUB... -l` in one go? Of course that would beg the question why can't one do `nodeos-sectl -d PUB1 -d PUB2` as well.

[the remark this answers is not in the document]
Comment (swatanabe-b1) [reply programs/nodeos-sectl/main.cpp]: > Should they? This wasn't clearly obvious to me. Would someone ever want to do `nodeos-sectl -d PUB... -l` in one go?

Is there any reason that this is better than running two separate commands?  Also, if you did that, the semantics are a bit confusing.  For instance, what does `nodeos-sectl -l -d PUB` do?

File: libraries/se-helpers/se-helpers.cpp [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+      fc::ec_key key = EC_KEY_new_by_curve_name(NID_X9_62_prime256v1);
+      *compact_sig = fc::crypto::r1::signature_from_ecdsa(key,my->pub_key._storage.get<fc::crypto::r1::public_key_shim>()._data, sig, digest);
+   } catch(...) {
+      CFRelease(signature);
+      CFRelease(digestData);
+      throw;
+   }
+
+   CFRelease(signature);
+   CFRelease(digestData);
```
[the remark this answers is not in the document]
Comment (spoonincode) [reply libraries/se-helpers/se-helpers.cpp]: I've been reluctant to go whole hog and create CoreFoundation wrappers for all these types; seemed a little overkill for such little usage. (Locally to experiment) I changed things around to use `fc::make_scoped_exit` which resolves the pathological leaks but I'm not really sure I'm pleased with the final appearance using that.

[the remark this answers is not in the document]
Comment (swatanabe-b1) [reply libraries/se-helpers/se-helpers.cpp]: unique_ptr with a deleter should work.

File: libraries/se-helpers/se-helpers.cpp [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+
+#include <fc/fwd_impl.hpp>
+
+#include <Security/Security.h>
+
+namespace eosio::secure_enclave {
+
+static std::string string_for_cferror(CFErrorRef error) {
+   CFStringRef errorString = CFCopyDescription(error);
+   char buff[CFStringGetLength(errorString) + 1];
```
[the remark this answers is not in the document]
Comment (spoonincode) [reply libraries/se-helpers/se-helpers.cpp]: Not something I'd lose sleep over, but easy enough to just dump the contents straight in to the std::string

File: tests/se_tests/se_tests.cpp [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+static bool pubkey_in_se_keyset(const fc::crypto::public_key& pub, const std::set<eosio::secure_enclave::secure_enclave_key>& keys) {
+   for(const eosio::secure_enclave::secure_enclave_key& sekey : keys)
+      if(sekey.public_key() == pub)
+         return true;
+
+   return false;
+}
+
+BOOST_TEST_GLOBAL_FIXTURE(se_cleanup_fixture);
```
[the remark this answers is not in the document]
Comment (swatanabe-b1) [reply tests/se_tests/se_tests.cpp:38]: This isn't a serious problem because se_tests is built as a separate executable.

**Pull request merged**